Add ability to start websocket connection with a frame size or a Codec

sazid opened the pull request:
<!-- Thanks for considering contributing actix! -->
<!-- Please fill out the following to make our reviews easy. -->

## PR Type
<!-- What kind of change does this PR make? -->
<!-- Bug Fix / Feature / Refactor / Code Style / Other -->
Feature


## PR Checklist

<!-- For draft PRs check the boxes as you complete them. -->

- [ ] Tests for the changes have been added / updated.
- [x] Documentation comments have been added / updated.
- [x] A changelog entry has been made for the appropriate packages.
- [x] Format code with the latest stable rustfmt


## Overview
<!-- Describe the current and new behavior. -->
<!-- Emphasize any breaking changes. -->
**New**: You can now start a websocket session with a frame size or a `actix_http::ws::Codec` by calling either `ws::start_with_frame_size()` or `ws::start_with_codec()`.

<!-- If this PR fixes or closes an issue, reference it here. -->
<!-- Closes #000 -->
Fixes #1873

sazid committed:
add fn's to start ws with frame size and codec

sazid committed:
add fn's to start ws with frame size and codec

sazid committed:
update changelog

sazid committed:
update changelog

robjtede commented on actix-web-actors/src/ws.rs:
```diff
@@ -40,6 +40,43 @@ where
+}
+
+/// Perform WebSocket handshake and start actor with the given
+/// `actix_http::ws::Codec`.
```
use intra-doc links

```suggestion
/// Perform WebSocket handshake and start actor with the given [`Codec`].
```

sazid committed:
Use intra-doc links.

Co-authored-by: Rob Ede <robjtede@icloud.com>

sazid committed:
Merge branch 'master' into feat/ws-codec-frame-size

robjtede committed:
major cleanup of middleware module (#1875)

* major cleanup of middleware module

* update changelog

fakeshadow committed:
refactor app_service (#1877)

fakeshadow committed:
More refactor of app_service (#1879)

fakeshadow committed:
replace cloneable service with httpflow abstraction (#1876)

robjtede committed:
add docs to recent additions

robjtede committed:
add http3 variant to protocol enum

robjtede committed:
fmt

fakeshadow committed:
fix bug where upgrade future is not reset properly (#1880)

robjtede committed:
address clippy lints

robjtede committed:
use env_logger builders in examples

robjtede committed:
prepare v4 beta releases (#1881)

fakeshadow committed:
fix extra branch in h1 dispatcher timer (#1882)

fakeshadow committed:
remove copyless dependency (#1884)

robjtede committed:
don't run awc doctests that rely on external public endpoints (#1888)

fakeshadow committed:
use sync method on Responder trait (#1891)

fakeshadow committed:
refactor Resource (#1883)

robjtede committed:
big clean up and docs improvmenet of types mod (#1894)

fakeshadow committed:
remove leaked box in REQUEST_POOL and RESPONSE_POOL (#1896)

fakeshadow committed:
remove actix-threadpool.use actix_rt::task::spawn_blocking (#1878)

fakeshadow committed:
refactor Scope (#1895)

fakeshadow committed:
use a non leak pool for HttpRequestInner (#1889)

Co-authored-by: Rob Ede <robjtede@icloud.com>

fakeshadow committed:
simplify h1 dispatcher (#1899)

Co-authored-by: Rob Ede <robjtede@icloud.com>

fakeshadow committed:
Refactor/service request (#1893)

fakeshadow committed:
fix actix-files doc about thread pool (#1898)

rschoon committed:
actix-files: Fix If-(Un)Modified to not consider sub-seconds (#1887)

fakeshadow committed:
Refactor h1 encoder (#1900)

therealprof reviewed (commented):
Nice. But since we're touching this it would be great to also add the possibility to interact with the handshake. Up to now I had to do the WS handshake myself a couple of times, one reason being the insufficient `max_size` (as addressed by this PR) and the other being the abilitity to introspect and set headers as well as dealing with WebSocket extensions.

robjtede commented:
We've had a chat on Gitter and are looking at a builder style solution to cover more of the custom cases. Either way this feature will be in v4.

therealprof commented:
> We've had a chat on Gitter and are looking at a builder style solution to cover more of the custom cases. Either way this feature will be in v4.

Excellent, maybe we should close this PR then to avoid confusion?

robjtede commented:
> Excellent, maybe we should close this PR then to avoid confusion?

I think @sazid is gonna tackle it in this PR. Correct me if I'm wrong. But I will mark as draft for now.

therealprof commented:
Marking as draft would have been my next suggestion, thanks. 😅

fakeshadow committed:
optimize ErrorHandler middleware (#1902)

sazid commented:
Oops, I think I didn't read the last two comments.

robjtede committed:
response header rework (#1869)

robjtede committed:
return option item from Extensions::insert (#1904)

robjtede committed:
Update PULL_REQUEST_TEMPLATE.md

robjtede committed:
cleanup top level doc comments

robjtede committed:
implement `App::data` as `App::app_data(Data::new(T)))` (#1906)

fakeshadow committed:
Refactor poll_keepalive for readability (#1901)

robjtede committed:
revive commented out tests (#1912)

lberrymage committed:
Remove HttpResponseBuilder::json2() (#1903)

It's not necessary to keep both json() and json2() around since the
former reduces the ownership of its parameter to a borrow only to pass
the reference to the latter. Users can instead borrow themselves when
passing an owned value: there doesn't need to be two separate functions.

This change also makes HttpResponseBuilder::json() take T: Deref so it
can accept both references and web extractors like web::Json.

sazid committed:
add WsSessionBuilder for building ws sessions

sazid committed:
Merge branch 'feat/ws-codec-frame-size' of https://github.com/sazid/actix-web into feat/ws-codec-frame-size

sazid committed:
return `Self` instead of `&mut Self` in WsResponseBuilder

sazid committed:
add tests for WsResponseBuilder

sazid commented:
Reopened a new PR with builder approach. 
https://github.com/actix/actix-web/pull/1920
